File: ios/Sources/FlowbizOnsite/TaskScheduler.swift
import Foundation

/// Cancellable handle for a scheduled task. Cancelling twice is harmless.
protocol ScheduledHandle {
    func cancel()
}

/// The SDK's serial execution seam (SPEC §1: GCD). Everything the transport
/// layer does — queue I/O, HTTP, backoff retries, heartbeat ticks — runs
/// through one implementation backed by a **serial** `DispatchQueue`, so
/// queue and flush state are thread-confined without locking. Injected so
/// tests drive time and execution manually.
protocol TaskScheduler {
    /// Runs `task` on the serial queue as soon as possible.
    func execute(_ task: @escaping @Sendable () -> Void)

    /// Runs `task` on the serial queue after `afterMillis`.
    func schedule(afterMillis: Int64, _ task: @escaping @Sendable () -> Void) -> any ScheduledHandle

    /// Runs `task` on the serial queue every `intervalMillis`, first fire
    /// one full interval after scheduling.
    func scheduleRepeating(intervalMillis: Int64, _ task: @escaping @Sendable () -> Void) -> any ScheduledHandle
}

/// Production `TaskScheduler` over a serial `DispatchQueue` (Slice 4 owns
/// its creation/lifecycle). One-shot delays use `DispatchWorkItem` +
/// `asyncAfter`; the repeating heartbeat uses a `DispatchSourceTimer` on the
/// same queue (SPEC §8 timer choice).
final class DispatchTaskScheduler: TaskScheduler, @unchecked Sendable {

    private final class WorkItemHandle: ScheduledHandle, @unchecked Sendable {
        private let item: DispatchWorkItem
        init(_ item: DispatchWorkItem) { self.item = item }
        func cancel() { item.cancel() }
    }

    private final class TimerHandle: ScheduledHandle, @unchecked Sendable {
        private let timer: DispatchSourceTimer
        init(_ timer: DispatchSourceTimer) { self.timer = timer }
        func cancel() { timer.cancel() }
    }

    private let queue: DispatchQueue

    /// - Parameter queue: the SDK's **serial** queue.
    init(queue: DispatchQueue) {
        self.queue = queue
    }

    func execute(_ task: @escaping @Sendable () -> Void) {
        queue.async(execute: task)
    }

    func schedule(afterMillis: Int64, _ task: @escaping @Sendable () -> Void) -> any ScheduledHandle {
        let item = DispatchWorkItem(block: task)
        queue.asyncAfter(deadline: .now() + .milliseconds(Int(afterMillis)), execute: item)
        return WorkItemHandle(item)
    }

    func scheduleRepeating(intervalMillis: Int64, _ task: @escaping @Sendable () -> Void) -> any ScheduledHandle {
        let timer = DispatchSource.makeTimerSource(queue: queue)
        let interval = DispatchTimeInterval.milliseconds(Int(intervalMillis))
        timer.schedule(deadline: .now() + interval, repeating: interval)
        timer.setEventHandler(handler: task)
        timer.resume()
        return TimerHandle(timer)
    }
}

File: android/sdk/src/main/kotlin/com/flowbiz/onsite/TaskScheduler.kt
package com.flowbiz.onsite

import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/** Cancellable handle for a scheduled task. Cancelling twice is harmless. */
internal interface ScheduledHandle {
    fun cancel()
}

/**
 * The SDK's serial execution seam (SPEC §1: single background
 * `ExecutorService`). Everything the transport layer does — queue I/O,
 * HTTP, backoff retries, heartbeat ticks — runs through one implementation
 * of this interface backed by a **single-threaded** scheduled executor, so
 * queue and flush state are thread-confined without locking. Injected so
 * tests drive time and execution manually.
 */
internal interface TaskScheduler {
    /** Runs [task] on the serial thread as soon as possible. */
    fun execute(task: Runnable)

    /** Runs [task] on the serial thread after [delayMillis]. */
    fun schedule(delayMillis: Long, task: Runnable): ScheduledHandle

    /**
     * Runs [task] on the serial thread every [intervalMillis], first fire
     * one full interval after scheduling (fixed delay between runs).
     */
    fun scheduleRepeating(intervalMillis: Long, task: Runnable): ScheduledHandle
}

/**
 * Production [TaskScheduler] over a single-threaded
 * [ScheduledExecutorService] (Slice 4 owns its creation/lifecycle).
 * Never throws: a rejected task (executor shut down) degrades to a no-op
 * handle (SPEC §3).
 */
internal class ExecutorTaskScheduler(
    private val executor: ScheduledExecutorService,
) : TaskScheduler {

    private class FutureHandle(private val future: ScheduledFuture<*>) : ScheduledHandle {
        override fun cancel() {
            try {
                future.cancel(false)
            } catch (_: Throwable) {
            }
        }
    }

    private object NoopHandle : ScheduledHandle {
        override fun cancel() {}
    }

    override fun execute(task: Runnable) {
        try {
            executor.execute(task)
        } catch (_: RejectedExecutionException) {
            SdkLog.debug("scheduler rejected task (shutting down)")
        }
    }

    override fun schedule(delayMillis: Long, task: Runnable): ScheduledHandle = try {
        FutureHandle(executor.schedule(task, delayMillis, TimeUnit.MILLISECONDS))
    } catch (_: RejectedExecutionException) {
        SdkLog.debug("scheduler rejected delayed task (shutting down)")
        NoopHandle
    }

    override fun scheduleRepeating(intervalMillis: Long, task: Runnable): ScheduledHandle = try {
        FutureHandle(
            executor.scheduleWithFixedDelay(task, intervalMillis, intervalMillis, TimeUnit.MILLISECONDS)
        )
    } catch (_: RejectedExecutionException) {
        SdkLog.debug("scheduler rejected repeating task (shutting down)")
        NoopHandle
    }
}
